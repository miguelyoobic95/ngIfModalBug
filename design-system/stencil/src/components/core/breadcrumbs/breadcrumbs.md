---
name: Breadcrumbs
category: Core Components
---

Breadcrumbs are a navigation pattern that can be implemented with `<yoo-breadcrumbs>`. They are always horizontal and if they have more than 7 items,
these will be collapsed into a context-menu starting with the root item at the bottom.

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`items`|string[ ]| required*   |Hierarchy of pages to present in the breadcrumbs - the last item will be highlighted|

```yoo-breadcrumbs.html
    <yoo-breadcrumbs></yoo-breadcrumbs>
```

```yoo-breadcrumbs.js
    var comp = document.querySelector('yoo-breadcrumbs');
    comp.items = ["item 1", "item2", "item 3", "item 4", "item 5", "item 6"];
```

```yoo-breadcrumbs-max.html
    <yoo-breadcrumbs></yoo-breadcrumbs>
```

```yoo-breadcrumbs-max.js 
    var comp = document.querySelector('yoo-breadcrumbs');
    comp.items = ["item 1", "item2", "item 3", "item 4", "item 5", "item 6", "item 7"];
```

```yoo-breadcrumbs-collapse.html
    <yoo-breadcrumbs></yoo-breadcrumbs>
```

```yoo-breadcrumbs-collapse.js 
    var comp = document.querySelector('yoo-breadcrumbs');
    comp.items = ["item 1", "item2", "item 3", "item 4", "item 5", "item 6", "item 7", "item 8", "item 9", "item 10"];
```

## Events

|Attr|Description|
|---|---|
|`itemSelected`|Returns the item the user clicks on|
