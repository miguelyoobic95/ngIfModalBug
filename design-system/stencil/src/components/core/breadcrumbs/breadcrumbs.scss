

:host {
    .breadcrumb {
        align-items: center;
        line-height: 2;
        font-size: 0.875rem;
        padding: 0.375rem 0rem;

        .breadcrumb-item {
            align-self: center;
            color: cssvar(dark-60);
            padding-right: 0.75rem;
            
            span {
                position: relative;
                @include text-truncate();                
                padding-right: 0.75rem;
            }
            
            &.active {
                color: cssvar(text-color);
            }

            &.more {
                span {
                    &.yo-right {
                        padding-right: 0rem;
                    }
                }
                .more-icons {
                    padding-left: 0.75rem; 
                    i {
                        &.yo-arrow-dropdown {
                            padding: 0rem 0.25rem;
                        }
                    }
                }
            }
        }

        yoo-context-menu {
            .context-container {
                flex-direction: column-reverse;                
                span {
                    padding: 0.5rem 1rem;
                    line-height: 1.5;
                    width: 100%;
                    position: relative;
                    border-top: 1px solid cssvar(dark-20);
                    &:last-child {
                        border-top: none;
                    }
                    &:hover {
                        background: cssvar(accent-05);
                        color: cssvar(dark);
                    }
                }
            }
        }
    }
}
