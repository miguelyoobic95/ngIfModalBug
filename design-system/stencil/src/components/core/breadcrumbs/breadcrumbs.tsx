import { Component, Prop, State, Event, EventEmitter, Element } from '@stencil/core';
import { getElementDimensions } from '../../../utils/helpers';
@Component({
    tag: 'yoo-breadcrumbs',
    styleUrl: 'breadcrumbs.scss',
    scoped: true,
    // tslint:disable-next-line:use-host-property-decorator
    host: {
        role: 'navigation'
    }
})
export class YooBreadcrumbsComponent {

    ITEM_WIDTH = 80;
    MAX_VISIBLE_ITEMS: number = 7;

    @Prop() items: string[] = [];

    @Event() itemSelected: EventEmitter<string>;

    @State() visibleItems = 7;

    @Element() host: HTMLStencilElement;

    componentWillLoad() {
        this.setItemNumber();
        // check the body width here and set max_steps accordingly
        window.addEventListener('resize', () => this.setItemNumber());
    }

    setItemNumber(): void {
        let width = getElementDimensions(this.host.parentElement).width;

        this.visibleItems = Math.min(Math.floor(width / this.ITEM_WIDTH), this.MAX_VISIBLE_ITEMS);
    }
    selectItem(item: string): void {
        this.itemSelected.emit(item);
    }

    isLastItem(index: number, arr: string[]): boolean {
        return index === arr.length - 1;
    }

    renderDefaultBreadcrumbItem(item: string, index: number, arr: string[]): JSX.Element {
        // last item is active
        return (
            <div class={'breadcrumb-item ' + (this.isLastItem(index, arr) ? 'active' : '')} onClick={this.selectItem.bind(this, item)}>
                <span>{item}</span>
                {(!this.isLastItem(index, arr) ? <i class="yo-right"></i> : '')}
            </div>
        );
    }

    renderCollapsedBreadcrumbItem(item: string): JSX.Element {
        return (<span onClick={this.selectItem.bind(this, item)}>{item}</span>);
    }

    // totalItems > MAX_VISIBLE_ITEMS creates a dropdown
    render(): JSX.Element {

        let collapsedItems = [];
        if (this.items.length > this.MAX_VISIBLE_ITEMS) {
            collapsedItems = this.items.slice(0, this.items.length - this.visibleItems);
        }
        let visibleItems;
        collapsedItems.length > 0 ? visibleItems = this.items.slice(this.items.length - this.visibleItems) : visibleItems = this.items;

        return (
            <div class={'breadcrumb ' + (collapsedItems.length > 0 ? 'long' : '')} attr-layout="row">
            {collapsedItems.length > 0 ?
                <yoo-context-menu>
                    <div slot="trigger" class="breadcrumb-item more">
                        <span class="more-icons">
                            <i class="yo-more"></i> <i class="yo-arrow-dropdown"></i>
                        </span>
                        <span class="yo-right"></span>
                    </div>
                    <div class="context-container" attr-layout="column">
                        {collapsedItems.map(item => this.renderCollapsedBreadcrumbItem(item))}
                    </div>
                </yoo-context-menu>
            : ''}
            {visibleItems.map((item, index, arr) => this.renderDefaultBreadcrumbItem(item, index, arr))}
            </div>
        );
    }
}
