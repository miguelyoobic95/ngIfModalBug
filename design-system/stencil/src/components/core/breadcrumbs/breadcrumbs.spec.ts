import { TestWindow } from '@stencil/core/testing';
let mockDimensions = jest.fn();
jest.mock('../../../utils/helpers', () => {
    return {
        getElementDimensions: mockDimensions
    };
});
import { YooBreadcrumbsComponent } from './breadcrumbs';

describe('YooBreadcrumbsComponent', () => {
    it('should build', () => {
        expect(new YooBreadcrumbsComponent()).toBeTruthy();
    });

    describe('property rendering', () => {
        beforeEach(() => {
            mockDimensions.mockImplementation(() => {
                return {width: 600, height: 100};
            });
        });
        it('Should have navigation items', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooBreadcrumbsComponent],
                html: '<yoo-breadcrumbs></yoo-breadcrumbs>'
            });
            element.items = ['Item 1', 'Item 2'];
            await window.flush();
            expect(element).toMatchSnapshot();
            expect(element.querySelectorAll('.breadcrumb-item')).toHaveLength(2);
        });
        it('Should have an active navigation item', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooBreadcrumbsComponent],
                html: '<yoo-breadcrumbs></yoo-breadcrumbs>'
            });
            element.items = ['Item 1', 'Item 2'];
            await window.flush();
            expect(element).toMatchSnapshot();
            expect(element.querySelectorAll('.active')).toHaveLength(1);
            expect(element.querySelector('.active').querySelector('span').innerHTML).toBe('Item 2');
        });
        it('Should have a the correct number of items', async () => {
            mockDimensions.mockImplementation(() => {
                return {width: 400, height: 100};
            });
            let window = new TestWindow();
            let element = await window.load({
                components: [YooBreadcrumbsComponent],
                html: '<yoo-breadcrumbs></yoo-breadcrumbs>'
            });
            element.items = ['Item 1', 'Item 2', 'Item 3', 'Item 4', 'Item 5'];
            await window.flush();
            expect(element).toMatchSnapshot();
            expect(element.querySelectorAll('.breadcrumb-item')).toHaveLength(5);
        });
        it('Should have a context menu if there are more than 7 items navigation items', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooBreadcrumbsComponent],
                html: '<yoo-breadcrumbs></yoo-breadcrumbs>'
            });
            element.items = ['Item 1', 'Item 2', 'Item 3', 'Item 4', 'Item 5', 'Item 6', 'Item 7', 'Item 8'];
            await window.flush();
            expect(element).toMatchSnapshot();
            expect(element.querySelectorAll('yoo-context-menu')).toHaveLength(1);
            expect(element.querySelectorAll('.breadcrumb-item')).toHaveLength(8);
        });
        it('Should not have a context menu if there are less than 7 items navigation items', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooBreadcrumbsComponent],
                html: '<yoo-breadcrumbs></yoo-breadcrumbs>'
            });
            element.items = ['Item 1', 'Item 2', 'Item 3', 'Item 4', 'Item 5', 'Item 6'];
            await window.flush();
            expect(element).toMatchSnapshot();
            expect(element.querySelectorAll('yoo-context-menu')).toHaveLength(0);
            expect(element.querySelectorAll('.breadcrumb-item')).toHaveLength(6);
        });
        it('Should not have a context menu if there are exactly 7 items navigation items', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooBreadcrumbsComponent],
                html: '<yoo-breadcrumbs></yoo-breadcrumbs>'
            });
            element.items = ['Item 1', 'Item 2', 'Item 3', 'Item 4', 'Item 5', 'Item 6', 'Item 7'];
            await window.flush();
            expect(element).toMatchSnapshot();
            expect(element.querySelectorAll('yoo-context-menu')).toHaveLength(0);
            expect(element.querySelectorAll('.breadcrumb-item')).toHaveLength(7);
        });
    });

    describe('events', () => {
        it ('should emit on button click with rendering', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooBreadcrumbsComponent],
                html: '<yoo-breadcrumbs></yoo-breadcrumbs>'
            });
            element.items = ['Item 1', 'Item 2'];
            await window.flush();

            let res;
            let emitItem = jest.fn(ev => { res = ev.detail; });
            window.document.addEventListener('itemSelected', emitItem);
            await element.querySelector('.breadcrumb-item').click();
            expect(emitItem).toHaveBeenCalled();
            expect(res).toEqual('Item 1');
        });
    });
});
