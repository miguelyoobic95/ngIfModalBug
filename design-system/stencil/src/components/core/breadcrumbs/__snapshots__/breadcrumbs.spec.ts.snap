// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooBreadcrumbsComponent property rendering Should have a context menu if there are more than 7 items navigation items 1`] = `
<yoo-breadcrumbs
  class="hydrated"
  data-yoo-breadcrumbs-host=""
  role="navigation"
>
  
  <div
    attr-layout="row"
    class="breadcrumb long"
    data-yoo-breadcrumbs=""
  >
    <yoo-context-menu
      data-yoo-breadcrumbs=""
    >
      <div
        class="breadcrumb-item more"
        data-yoo-breadcrumbs=""
        slot="trigger"
      >
        <span
          class="more-icons"
          data-yoo-breadcrumbs=""
        >
          <i
            class="yo-more"
            data-yoo-breadcrumbs=""
          />
           
          <i
            class="yo-arrow-dropdown"
            data-yoo-breadcrumbs=""
          />
        </span>
        <span
          class="yo-right"
          data-yoo-breadcrumbs=""
        />
      </div>
      <div
        attr-layout="column"
        class="context-container"
        data-yoo-breadcrumbs=""
      >
        <span
          data-yoo-breadcrumbs=""
        >
          Item 1
        </span>
      </div>
    </yoo-context-menu>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 2
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 3
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 4
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 5
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 6
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 7
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item active"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 8
      </span>
      
    </div>
  </div>
</yoo-breadcrumbs>
`;

exports[`YooBreadcrumbsComponent property rendering Should have a the correct number of items 1`] = `
<yoo-breadcrumbs
  class="hydrated"
  data-yoo-breadcrumbs-host=""
  role="navigation"
>
  
  <div
    attr-layout="row"
    class="breadcrumb"
    data-yoo-breadcrumbs=""
  >
    
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 1
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 2
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 3
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 4
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item active"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 5
      </span>
      
    </div>
  </div>
</yoo-breadcrumbs>
`;

exports[`YooBreadcrumbsComponent property rendering Should have an active navigation item 1`] = `
<yoo-breadcrumbs
  class="hydrated"
  data-yoo-breadcrumbs-host=""
  role="navigation"
>
  
  <div
    attr-layout="row"
    class="breadcrumb"
    data-yoo-breadcrumbs=""
  >
    
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 1
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item active"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 2
      </span>
      
    </div>
  </div>
</yoo-breadcrumbs>
`;

exports[`YooBreadcrumbsComponent property rendering Should have navigation items 1`] = `
<yoo-breadcrumbs
  class="hydrated"
  data-yoo-breadcrumbs-host=""
  role="navigation"
>
  
  <div
    attr-layout="row"
    class="breadcrumb"
    data-yoo-breadcrumbs=""
  >
    
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 1
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item active"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 2
      </span>
      
    </div>
  </div>
</yoo-breadcrumbs>
`;

exports[`YooBreadcrumbsComponent property rendering Should not have a context menu if there are exactly 7 items navigation items 1`] = `
<yoo-breadcrumbs
  class="hydrated"
  data-yoo-breadcrumbs-host=""
  role="navigation"
>
  
  <div
    attr-layout="row"
    class="breadcrumb"
    data-yoo-breadcrumbs=""
  >
    
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 1
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 2
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 3
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 4
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 5
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 6
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item active"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 7
      </span>
      
    </div>
  </div>
</yoo-breadcrumbs>
`;

exports[`YooBreadcrumbsComponent property rendering Should not have a context menu if there are less than 7 items navigation items 1`] = `
<yoo-breadcrumbs
  class="hydrated"
  data-yoo-breadcrumbs-host=""
  role="navigation"
>
  
  <div
    attr-layout="row"
    class="breadcrumb"
    data-yoo-breadcrumbs=""
  >
    
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 1
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 2
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 3
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 4
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 5
      </span>
      <i
        class="yo-right"
        data-yoo-breadcrumbs=""
      />
    </div>
    <div
      class="breadcrumb-item active"
      data-yoo-breadcrumbs=""
    >
      <span
        data-yoo-breadcrumbs=""
      >
        Item 6
      </span>
      
    </div>
  </div>
</yoo-breadcrumbs>
`;
